group/secp256k1/scalar.go

```diff
@@ -0,0 +1,195 @@
+	s.int().Mod(s.int(), groupOrder)
+	return s
+}
+
+func (s *secp256k1Scalar) String() string {
+	return fmt.Sprintf("scalar{%x}", (*big.Int)(s))
+}
+
+var scalarZero = zero()
+
+// Equal returns true if s and sPrime represent the same value modulo the group
+// order, false otherwise
+func (s *secp256k1Scalar) Equal(sPrime kyber.Scalar) bool {
+	difference := zero().Sub(s.int(), toInt(sPrime))
+	return scalarZero.Cmp(difference.Mod(difference, groupOrder)) == 0
+}
+
+// Set copies sPrime's value (modulo groupOrder) to s, and returns it
+func (s *secp256k1Scalar) Set(sPrime kyber.Scalar) kyber.Scalar {
+	return (*secp256k1Scalar)(s.int().Mod(toInt(sPrime), groupOrder))
+}
+
+// Clone returns a copy of s mod groupOrder
+func (s *secp256k1Scalar) Clone() kyber.Scalar {
+	return (*secp256k1Scalar)(zero().Mod(s.int(), groupOrder))
+}
+
+// SetInt64 returns s with value set to v modulo groupOrder
+func (s *secp256k1Scalar) SetInt64(v int64) kyber.Scalar {
+	return (*secp256k1Scalar)(s.int().SetInt64(v)) // NB: v < groupOrder
+}
+
+// Zero sets s to 0 mod groupOrder, and returns it
+func (s *secp256k1Scalar) Zero() kyber.Scalar {
+	return s.SetInt64(0)
+}
+
+// Add sets s to a+b mod groupOrder, and returns it
+func (s *secp256k1Scalar) Add(a, b kyber.Scalar) kyber.Scalar {
+	s.int().Add(toInt(a), toInt(b))
+	return s.modG()
+}
+
+// Sub sets s to a-b mod groupOrder, and returns it
+func (s *secp256k1Scalar) Sub(a, b kyber.Scalar) kyber.Scalar {
+	s.int().Sub(toInt(a), toInt(b))
+	return s.modG()
+}
+
+// Neg sets s to -s mod groupOrder, and returns it
+func (s *secp256k1Scalar) Neg(a kyber.Scalar) kyber.Scalar {
+	s.int().Neg(toInt(a))
+	return s.modG()
+}
+
+// One sets s to 1 mod groupOrder, and returns it
+func (s *secp256k1Scalar) One() kyber.Scalar {
+	return s.SetInt64(1)
+}
+
+// Mul sets s to a*b mod groupOrder, and returns it
+func (s *secp256k1Scalar) Mul(a, b kyber.Scalar) kyber.Scalar {
+	// TODO(alx): Make this faster
+	s.int().Mul(toInt(a), toInt(b))
+	return s.modG()
+}
+
+// Div sets s to a*b⁻¹ mod groupOrder, and returns it
+func (s *secp256k1Scalar) Div(a, b kyber.Scalar) kyber.Scalar {
+	if toInt(b).Cmp(scalarZero) == 0 {
+		panic("attempt to divide by zero")
+	}
+	// TODO(alx): Make this faster
+	s.int().Mul(toInt(a), zero().ModInverse(toInt(b), groupOrder))
+	return s.modG()
+}
+
+// Inv sets s to s⁻¹ mod groupOrder, and returns it
+func (s *secp256k1Scalar) Inv(a kyber.Scalar) kyber.Scalar {
+	if toInt(a).Cmp(scalarZero) == 0 {
+		panic("attempt to divide by zero")
+	}
+	s.int().ModInverse(toInt(a), groupOrder)
+	return s
+}
+
+// Pick sets s to a random value mod groupOrder sampled from rand, and returns
+// it
+func (s *secp256k1Scalar) Pick(rand cipher.Stream) kyber.Scalar {
+	return s.Set((*secp256k1Scalar)(random.Int(groupOrder, rand)))
+}
+
+// MarshalBinary returns the big-endian byte representation of s, or an error on
+// failure
+func (s *secp256k1Scalar) MarshalBinary() ([]byte, error) {
+	b := toInt(s.modG()).Bytes()
```

Comment:
Why is that modG in there? It has the consequence that a MarshalBinary can change the underlying data, which is surprising to callers. Compare to pairing/bn256/point.go, where MarshalBinary uses Clone when it knows it's about to call MakeAffine and modify the point.

Reply:
This is defensive programming, since a secp256k1 should already be reduced mod the groupOrder. I can take a clone first, though.